gssapi/MICToken.go

```diff
@@ -0,0 +1,195 @@
+func (mt *MICToken) ComputeAndSetChecksum(key types.EncryptionKey, keyUsage uint32) error {
+	if mt.Payload == nil {
+		return errors.New("payload has not been set")
+	}
+	if mt.Checksum != nil {
+		return errors.New("checksum has already been computed")
+	}
+	checksum, err := mt.ComputeChecksum(key, keyUsage)
+	if err != nil {
+		return err
+	}
+	mt.Checksum = checksum
+	return nil
+}
+
+// ComputeChecksum computes and returns the checksum of this token, computed using the passed key and key usage.
+// Confirms to RFC 4121 in that the checksum will be computed over { body | header }.
+// In the context of Kerberos MIC tokens, mostly keyusage GSSAPI_ACCEPTOR_SIGN (=23)
+// and GSSAPI_INITIATOR_SIGN (=25) will be used.
+// Note: This will NOT update the struct's Checksum field.
+func (mt *MICToken) ComputeChecksum(key types.EncryptionKey, keyUsage uint32) ([]byte, error) {
+	if mt.Payload == nil {
+		return nil, errors.New("cannot compute checksum with uninitialized payload")
+	}
+	checksumMe := make([]byte, MICHdrLen+len(mt.Payload))
+	copy(checksumMe[0:], mt.Payload)
+	copy(checksumMe[len(mt.Payload):], getMICChecksumHeader(mt.Flags, mt.SndSeqNum))
+
+	encType, err := crypto.GetEtype(key.KeyType)
+	if err != nil {
+		return nil, err
+	}
+	return encType.GetChecksumHash(key.KeyValue, checksumMe, keyUsage)
+}
+
+// Build a header suitable for a checksum computation
+func getMICChecksumHeader(flags byte, senderSeqNum uint64) []byte {
+	header := make([]byte, MICHdrLen)
+	copy(header[0:2], getGssMICTokenId()[:])
+	header[2] = flags
+	copy(header[3:8], fillerBytes()[:])
+	binary.BigEndian.PutUint64(header[8:16], senderSeqNum)
+	return header
+}
+
+// Verify Checksum computes the token's checksum with the provided key and usage,
+// and compares it to the checksum present in the token.
+// In case of any failure, (false, err) is returned, with err an explanatory error.
+func (mt *MICToken) VerifyChecksum(key types.EncryptionKey, keyUsage uint32) (bool, error) {
+	computed, err := mt.ComputeChecksum(key, keyUsage)
+	if err != nil {
+		return false, err
+	}
+	if !hmac.Equal(computed, mt.Checksum) {
+		return false, fmt.Errorf(
+			"checksum mismatch. Computed: %s, Contained in token: %s",
+			hex.EncodeToString(computed), hex.EncodeToString(mt.Checksum))
+	}
+	return true, nil
+}
+
+// Unmarshal bytes into the corresponding MICToken.
+// If expectFromAcceptor is true we expect the token to have been emitted by the gss acceptor,
+// and will check the according flag, returning an error if the token does not match the expectation.
+func (mt *MICToken) Unmarshal(b []byte, expectFromAcceptor bool) error {
+	if len(b) < MICHdrLen {
+		return errors.New("bytes shorter than hedaer length")
+	}
+	if !bytes.Equal(getGssMICTokenId()[:], b[0:2]) {
+		return fmt.Errorf("wrong Token ID, Expected %s, was %s",
+			hex.EncodeToString(getGssMICTokenId()[:]),
+			hex.EncodeToString(b[0:2]))
+	}
+	flags := b[2]
+	isFromAcceptor := flags&0x01 == 1
```

Comment:
Would be good to create constants for the flag values and use the constant in the comparison.
The values are here: https://tools.ietf.org/html/rfc4121#section-4.2.2
So I suggest:
```
const (
    MICTokenFlagSentByAcceptor = 0
    MICTokenFlagSealed = 1
    MICTokenFlagAcceptorSubkey = 2
)
```
therefore...
```
isFromAcceptor := flags&0x01 == MICTokenFlagSealed
```